Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -64,21 +65,29 @@ final class ConnectionPool {
     /// When the pool is asked for a new connection, it creates a `Key` from the url associated to the `request`. This key
     /// is used to determine if there already exists an associated `HTTP1ConnectionProvider` in `providers`.
     /// If there is, the connection provider then takes care of leasing a new connection. If a connection provider doesn't exist, it is created.
-    func getConnection(for request: HTTPClient.Request, preference: HTTPClient.EventLoopPreference, on eventLoop: EventLoop, deadline: NIODeadline?, setupComplete: EventLoopFuture<Void>) -> EventLoopFuture<Connection> {
+    func getConnection(_ request: HTTPClient.Request,
+                       preference: HTTPClient.EventLoopPreference,
+                       taskEventLoop: EventLoop,
+                       deadline: NIODeadline?,
+                       setupComplete: EventLoopFuture<Void>,
+                       logger: Logger) -> EventLoopFuture<Connection> {
```

Comment:
Yeah passing it around sounds okey (the only thing we really can do tbh to preserve metadata).

There's two things to keep in mind, though I'm okey with punting on them for now.

- once we have a "context thing" we're somewhat likely to want to break those APIs and pass around context, rather than the logger. Context would be able to offer a logger though. Are we ok with breaking APIs then, or would we "add them" next to these and deprecate these...?
- I'm not sold on the "context" as _last_ parameter and would want to eventually talk about it being the _first_ parameter... Both have downsides, let's shelf this discussion for now, but as a heads up. More interested in knowing how we'd change those APIs once we have context. ( https://golang.org/pkg/context/ and systems like akka etc... one kind of learns to just ignore the "context" param when reading methods, but the trailing one is a bit weird thb...) anyway, let's bikeshed that at some point I guess -- as we'd want all APIs be consistent
  - I guess "last" in those APIs since you want it to be optional... anyway, long bikeshed topic when we'll have context objects and PoCs ready 😉

Reply:
@ktoso yeah, we've discussed first vs. last before, internally, a long time ago.

Here's why we landed on _last_: Swift's APIs are supposed to be fluent. So for example you want to write

```swift
let foreground = Color(red: 32, green: 64, blue: 128)
let newPart = factory.makeWidget(gears: 42, spindles: 14)
let ref = Link(target: destination)
```

And it's supposed to read like

> Factory, make a widget with 42 gears and 14 spindles.

If we now add a logger or a context into the picture, then we can either

```swift
let newPart = factory.makeWidget(context: context, gears: 42, spindles: 14)
```

which will become

> Factory, make a widget with context, 42 gears, and 14 spindles.

or

```swift
let newPart = factory.makeWidget(gears: 42, spindles: 14, context: context)
```

> Factory, make a widget with 42 gears, 14 spindles, and context.

I think arguably, reading out "context" or "logger" should just be skipped because it's _less important_ to decide what'll be going on. Therefore, I really think the best place is at the very end.

Reply:
> @ktoso yeah, we've discussed first vs. last before, internally, a long time ago.

I know, and I challenge us to rethink this before we enforce guidelines to the wider ecosystem, or at least well-define all the cases and and rationale for the various scenarios.

Notably, the "context last" is _not a consistent guideline_ IMHO and "consistency is king" in any such guidelines.

 E.g. if you're any library which offers context to users, today those look like this:

```swift
func channelRead(context: ChannelHandlerContext, data: NIOAny) ...
func readHandshakeResponse(context: ChannelHandlerContext, bytes: InboundIn) ...
// or 
... = .handleThings { context, things in ... }
func onThingsArrived(context: MyFramework.Context, things: Things) {
  // context.baggage // trace context / baggage context
  // context.log // logger with pre populated trace / baggage metadata
```

so it's natural when I delegate to my functions to carry like 

```swift
 // context, thingy
let x = factory.makeWidget(context, thingy: thingy)
makeWidget(context, x: x)
```

yet the guideline of "context last because it does not matter" goes against this.

I find it to be inconsistent and fatiguing if one has to constantly mentally flip around the parameter order if "library" or "my code" is concerned.

---

Minor nitpick about the spelling of the snippet:

```swift
let newPart = factory.makeWidget(context: context, gears: 42, spindles: 14)
```

is spelled very verbosely, for APIs intended to be called by library users / their own code, I would expect it spelled as 


```swift
let newPart = factory.makeWidget(context, gears: 42, spindles: 14)
```

which makes it shorter (and we already talk about "learn to not read out loud context).

--- 

The argument for "context last" goes as you say:

> I think arguably, **reading out "context" or "logger" should just be skipped** because it's less important to decide what'll be going on. Therefore, I really think the best place is at the very end.

so we learn/encourage to not read it out loud _anyway_ (!). 

Putting it last however also means it is easier to _miss_ propagating somewhere -- which is a _huge_ problem if we end up with baggage context (trace context) - "dropped" traces are a huge pain in the neck.

I can sympathise with the "context is ugly, let's put it in the end" however the confusion around "but this type of context goes first, but the other context type goes last" adds a lot of mental overhead when building APIs which I don't think is helping arrive at clean and consistent APIs throughout user code and library code.

Reply:
Also to keep in mind... in this specific library / example you argue to use `lastParam: Logger = noop` implying that if you forget to pass one "it's fine" and it really is low priority and some optional thing one may want to pass sometimes -- in that context it being last sounds fine.

However, if we want to get to a world where we don't accidentally "drop" traces and metadata, passing the context is _not_ optional - whenever you forget to pass around a context you _broke_ tracing. So while "reading it (context) out loud" definitely is not something we care about, "making sure it is being carried around" is super important, as is taking the mental overhead off people wether they use some library or their code and "where" the context should be.

I also understand a lot of this is "because we're used to X" but I'd really want to discuss the tradeoffs and consistency aspects on more examples, before we make a strong recommendation (esp. if we make a guideline that a lot of APIs end up ignoring).

Reply:
fwiw, I am also in the context comes first camp, for many of the same reasons @ktoso highlights

Reply:
@ktoso In NIO we just got it wrong IMHO. We didn't fix it in NIO 2 because that would've made a hard migration even harder (we just changed `ctx` to `context` which can be done easily with `sed`).

I used to also be in the "context goes first" camp but the only argument in favour of first is consistency. I don't think consistency with NIO (which IMHO got it wrong) is the right choice here.

Swift has many APIs that read like `addObserver<O: Observer>(_ observer: O)`. If we added `context` as the first parameter, that becomes a huge mess:

- `addObserver<O: Observer>(context: Context, _ observer: O)` is just wrong
- `addObserver<O: Observer>(context: Context, observer: O)` is less bad but much longer than necessary
- `addObserver<O: Observer>(_ observer: O, context: Context)` has exactly the right feel in Swift

More importantly, you're only supposed to add the "Observer" bit when you're "compensating for weak type information". So in many APIs where it's much clearer, you're supposed to just name the method `add`.

Example: `add(_ rectangle: Rectangle)`. If we now add `context`, we end up in a very bad spot:

- `add(context: Context, _ rectangle: Rectangle)` is again just wrong because we're adding a rectangle and not a context
- `add(context: Context, rectangle: Rectangle)` is less bad but super confusing
- `add(_ rectangle: Rectangle, context: Context)` is again as good as it gets.

Basically, you cannot just skip over `context` when it's at the front but it works very well if at the back.

Reply:
CC @airspeedswift for API advise.

Reply:
Right, it's all about consistency (which is 👑). Since today is not really upheld in the libs that's where my problem lies - less about "first vs last", to me at least. As we're "early enough" perhaps it's the right time to make the rule and enforce it...

So, in order to make the consistency/👑 happy... would you say we try all new (server at least) APIs stick to the "last" rule, and if automated tooling to migrate existed, would NIO follow that perhaps as well (more below)?

I'm happy to discover that:
- gRPC sticks to the last rule already, so that's good: https://github.com/grpc/grpc-swift/blob/3109a71355f84881887957cfdabbac41150e0793/Sources/Examples/HelloWorld/Model/helloworld.grpc.swift#L66 we're able to make it such in http-client and 
- RediStack does not seem to do context things in its APIs yet
- HTTPClient we're doing now and can do the "last"
- Vapor does not seem to have such thing?

Since I'm mostly concerned about the medium to high level libs all doing the same (which are vetted by the SSWG) -- that's where most of the context"passing around" is happening -- and where most of the mix of various libraries exists if we're able to pick a rule and stick to it that's the most important to me. 

I'd be okey with sweeping NIO's current context-first under the rug, **BUT** we should be explicit about it when we propose the strict guideline about passing around contexts, and it'd be good that NIO addressess this when able to (which means years+ probably, but still worth aligning all server APIs on this IMO, that's what Go does right -- "same style" for these things). 

With swift-syntax I believe we can easily provide a reformatter that would shift around parameters in signatures in end user's code.


--- 

So... 

- existing and ongoing development other libs should switch to "context last" (oh boy, I have 1,000s of places to change now, oh boy 🤣)
- our work in the GSoC context lib should assume and show in examples "context last"
- I'm kind of betting on NIO and related libs could follow suit _when they're able to_ (we're taking multi year here, no rush, but intent matters -- it's annoying to have a "rule" with core libs being exceptions)

WDYT?

Reply:
An important thing here also: It is `context: Context` and NOT `Context?` - this is _hugely_ important.
The optional around `Logger?` in this proposal had me quite worried -- dropping a context is disastrous, so we'd not want it be optional and default to nothing (like we do here for the logger).

Reply:
@ktoso I agree that consistency is important and if we could redo NIO, I'm pretty sure we'd do it last. Said that, NIO's `ChannelHandlerContext` isn't quite the same as a `Context`. In almost all `ChannelHandler` method implementation you will have to use `context`, or else your program will just drop events, so it won't work correctly.

Regarding `Optional<Logger>` and `Optional<Context>`: I _think_ non-optional is better here, possibly defaulted to some `.disableLogging` / `Context.empty` / `Context()`? But maybe not, can always have the user type `Context()` to pass an empty one. If they already got one, they'll just pass it along which is the right thing.

I don't have a strong opinion yet I think however `Optional<Thing>` can be very dangerous in API design, example [here](https://github.com/swift-server/async-http-client/issues/197). You need to be very mindful that the user will be (by Xcode) offered a `.none` completion. This can be very misleading, for example in AsyncHTTPClient's redirect configuration `redirects: .none` means "default redirects", ie. 5 redirects max and _not_ "no redirects". `redirects: .disallow` is what makes you have "no redirects". Both of those wouldn't be an issue with `Optional<Logger>` and `Optional<Context>` because `.none` gives you the right idea but we should still be mindful I think.

Reply:
Ok two topics here :) 

> @ktoso I agree that consistency is important and if we could redo NIO, I'm pretty sure we'd do it last. 
> 
> Said that, NIO's ChannelHandlerContext isn't quite the same as a Context. In almost all ChannelHandler method implementation you will have to use context, or else your program will just drop events, so it won't work correctly.

I had a long chat about with Johannes now about similar APIs to this but which are not quite as low level as NIO, and what we should do there. Overall consistency is what matters and if we can get most APIs to do "last", and NIO remains with it's historical difference I guess we can live with that -- what matters is that high level tools like http clients, grpc and others follow the same style -- and we have time _now_ to make it so. 

I do agree that last reads well in general... and poked around at some APIs where we had it first at how last would look like, seems managable.

--- 

#### Optional / default values for context

> Regarding Optional<Logger> and Optional<Context>: I think non-optional is better here,

Yes, absolutely. Optional is actively hurtful for such, especially baggage / trace contexts. The moment it becomes "easy" to drop the context, people _will_ even accidentally and if it's hard to spot where the drop happened then fixing those is a a pain as well. A dropped context / trace means the entire trace becomes dangling and useless.

> I don't have a strong opinion yet. I do think however Optional<Thing> can be very dangerous in API design,  [...]

> possibly defaulted to some .disableLogging / Context.empty / Context()? 

Please no default values for baggage contexts 😱 The logging case here for the time being I could swallow, since nothing "too bad" happens if you don't pass it -- things won't be logged, end of story. But as mentioned above, any time it becomes easy or even automatic to drop context it _will_ be dropped by someone, rendering the entire trace propagation useless (aborting a trace -- no end of a trace -> trace never shows up -> the search beings who/where/how dropped the trace in which library and how to fix/update the lib).

>  But maybe not, can always have the user type Context() to pass an empty one. If they already got one, they'll just pass it along which is the right thing.

Yes, exactly -- we have to make the right thing, the natural/simple thing. The natural and right thing is most of the time: I got some baggage and I just pass it around. Usually it's the web framework or something that always gives you some baggage context already anyway, so passing it around should not be too hard :)


---